File: dec-21/part1.py
if __name__ == '__main__':
    field = []
    with open('data.txt', 'r') as f:
        while True:
            line = f.readline()
            if not line:
                break
            field.append(list(line.strip()))

    rows, cols = len(field), len(field[0])
    start = None
    for row in range(rows):
        for col in range(cols):
            if field[row][col] == 'S':
                start = (row, col)
                break
    assert start

    field[start[0]][start[1]] = '.'
    reachable = {start}

    for _ in range(64):
        current, reachable = reachable, set()
        for row, col in current:
            candidates = [
                (row - 1, col),
                (row + 1, col),
                (row, col - 1),
                (row, col + 1),
            ]
            for candidate_row, candidate_col in candidates:
                if not (0 < candidate_row < rows and 0 <= candidate_col < cols):
                    continue
                if field[candidate_row][candidate_col] != '.':
                    continue
                reachable.add((candidate_row, candidate_col))

    print(len(reachable))

File: 2025/dec-04/part1.go
//go:build ignore

package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func main() {
	data, _ := os.ReadFile("data.txt")
	field := []string{}

	for _, line := range strings.Split(string(data), "\n") {
		if line == "" {
			continue
		}
		field = append(field, line)
	}

	rows := len(field)
	cols := len(field[0])

	reachable := map[[2]int]bool{}
	for row := 0; row < rows; row++ {
		for col := 0; col < cols; col++ {
			cell := string(field[row][col])

			if cell != "@" {
				continue
			}

			adjacentCount := 0
			adjacent := [][2]int{
				{row - 1, col - 1},
				{row - 1, col},
				{row - 1, col + 1},
				{row, col - 1},
				{row, col + 1},
				{row + 1, col - 1},
				{row + 1, col},
				{row + 1, col + 1},
			}

			for _, adj := range adjacent {
				arow := adj[0]
				acol := adj[1]
				if arow < 0 || arow >= rows || acol < 0 || acol >= cols {
					continue
				}
				acell := string(field[arow][acol])

				if acell == "@" {
					adjacentCount += 1
				}
			}

			if adjacentCount < 4 {
				reachable[[...]int{row, col}] = true
			}
		}
	}

	// print result field
	out, _ := os.Create("data.out")
	writer := bufio.NewWriter(out)

	for row := 0; row < rows; row++ {
		for col := 0; col < cols; col++ {
			cell := [2]int{row, col}
			if reachable[cell] {
				writer.WriteString("x")
			} else {
				writer.WriteString(string(field[row][col]))
			}
		}
		writer.WriteString("\n")
	}
	writer.Flush()
	out.Close()

	fmt.Println(len(reachable))
}
